tests/test_re.py

```diff
@@ -0,0 +1,520 @@
+        self.istop = istop
+
+    def _build_re(self):
+        return f"{self.base.build_re()}{{{self.istart},{self.istop}}}"
+
+    @staticmethod
+    def make_repetition(base, draw):
+        start = draw(st.integers(min_value=0, max_value=MAXREPEAT))
+        stop = draw(st.integers(min_value=start, max_value=MAXREPEAT))
+        return StartStop(base, start, stop, draw(st.booleans()))
+
+
+class Stop(Repetition):
+    istart = 0
+
+    def __init__(self, base, istop, lazy):
+        Repetition.__init__(self, base, lazy)
+        self.istop = istop
+
+    def _build_re(self):
+        return f"{self.base.build_re()}{{,{self.istop}}}"
+
+    @staticmethod
+    def make_repetition(base, draw):
+        stop = draw(st.integers(min_value=0, max_value=MAXREPEAT))
+        return Stop(base, stop, draw(st.booleans()))
+
+
+class Start(Repetition):
+    istop = MAXREPEAT
+
+    def __init__(self, base, istart, lazy):
+        Repetition.__init__(self, base, lazy)
+        self.istart = istart
+
+    def _build_re(self):
+        return f"{self.base.build_re()}{{{self.istart},}}"
+
+    @staticmethod
+    def make_repetition(base, draw):
+        start = draw(st.integers(min_value=0, max_value=MAXREPEAT))
+        return Start(base, start, draw(st.booleans()))
+
+
+class Sequence(RecursiveRe):
+    def __init__(self, bases):
+        self.bases = bases
+
+    def can_be_empty(self):
+        return all(base.can_be_empty() for base in self.bases)
+
+    def matching_string(self, draw, state):
+        return "".join(b.matching_string(draw, state) for b in self.bases)
+
+    def non_matching_string(self, draw, state):
+        if self.can_be_empty():
+            raise CantGenerateNonMatching
+        nonempty_positions = [
+            i for (i, b) in enumerate(self.bases) if not b.can_be_empty()
+        ]
+        res = []
+        for base_pos in nonempty_positions:
+            res.append(self.bases[base_pos].non_matching_string(draw, state))
+        return "".join(res)
+
+    def build_re(self):
+        return "".join(b.build_re() for b in self.bases)
+
+    @staticmethod
+    def make_with_base(base, draw):
+        return Sequence(
+            [
+                base(draw)
+                for i in range(draw(st.integers(min_value=2, max_value=MAXREPEAT)))
+            ]
+        )
+
+
+class SequenceWithBackref(Sequence):
+    @staticmethod
+    def make_with_base(base, draw):
+        bases = [
+            base(draw)
+            for i in range(draw(st.integers(min_value=2, max_value=MAXREPEAT)))
+        ]
+        if IS_PYPY and sys.pypy_version_info < (7, 3, 1):
+            # PyPy before 7.3.1 actually has broken group references!
+            return Sequence(bases)
+        # the following code would have found the bug in
+        # https://foss.heptapod.net/pypy/pypy/commit/c83c263f9f00d18d48ef536947c9b61ca53e01a2
+        group = draw(st.integers(min_value=0, max_value=len(bases) - 1))
+        name = gensym()
```

Comment:
You can use the [`st.shared()` strategy](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.shared) to manage shared state, like:

```python
used_names = draw(st.shared(st.builds(set), key="group names"))  # generates then caches a set
name = draw(st.text().filter(lambda s: s not in used_names)  # or integers().map("symbol{}".format), etc
used_names.add(name)
```

And tada, no more global state so examples will be truly reproducible.

Reply:
fixed this, thanks for teaching me this feature!